src/python/pants/conftest.py

```diff
@@ -0,0 +1,30 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pathlib import Path
+
+# The top-level `pants` module must be a namespace package, because we build two dists from it
+# (pantsbuild.pants, and pantsbuild.pants.testutil) and consumers of these dists need to be
+# able to import from both.
+#
+# In fact it is an *implicit* namespace package - that is, it has no __init__.py file.
+# See https://packaging.python.org/guides/packaging-namespace-packages/ .
+#
+# Unfortunately, the presence or absence of __init__.py affects how pytest determines the
+# module names for test code. For details see
+# https://docs.pytest.org/en/stable/goodpractices.html#test-package-name .
+#
+# Usually this doesn't matter, as tests don't typically care about their own module name.
+# But we have tests (notably those in src/python/pants/engine) that create @rules and
```

Comment:
Hrm, those tests seem dubious - especially if they are the only thing tripping this all up. If its isolated to src/python/pants/engine/rules_test.py in particular, that might be good to add to the note here to point the way to undoing this robustly if possible.

Reply:
I'm not sure they're dubious. They expect rules named `pants.foo.bar.func_name`, which is what "production" rule names would be, and they will now see `foo.bar.func_name` instead. In some cases the functionality they are testing is rule name generation, and in other cases the rule names are embedded in error messages they test for, that kind of thing. But these all seem like legit cases where the test needs to know its own module name. 

Of course we could update the tests to expect these other module names instead, but that's also weird - we're deforming the tests because of a pytest quirk.

One thing we could maybe do is have the rules used by these tests live in a separate, non-test file that the test imports. Those files aren't subject to test discovery and are just imported the normal way, so they would have the right name. It's not ideal, but none of this is.